pkg/kubelet/kubelet.go

```
@@ -2276,32 +2277,73 @@ func (kl *Kubelet) rejectPod(pod *api.Pod, reason, message string) {
 }
 
+// getNodeAnyWay() must return a *api.Node which is required by RunGeneralPredicates().
+// The *api.Node is obtained as follows:
+// 1. When kubelet is in standaloneMode, manufactures *api.Node with kubelet
+// 2. Try to get *api.Node from kubelet's nodeInfo, if failed, we try 3.
+// 3. Try to get *api.Node from apiserver
+func (kl *Kubelet) getNodeAnyWay() (*api.Node, error) {
+	var node *api.Node
+	if kl.standaloneMode {
+		if n, err := kl.initialNodeStatus(); err == nil {
+			return n, nil
+		}
+	}
+	if n, err := kl.nodeInfo.GetNodeInfo(kl.nodeName); err == nil {
+		return n, nil
+	}
+	node, err := kl.kubeClient.Core().Nodes().Get(kl.nodeName)
+	if err != nil {
+		return node, err
+	}
+	return node, nil
+}
+
 // canAdmitPod determines if a pod can be admitted, and gives a reason if it
 // cannot. "pod" is new pod, while "pods" include all admitted pods plus the
 // new pod. The function returns a boolean value indicating whether the pod
 // can be admitted, a brief single-word reason and a message explaining why
 // the pod cannot be admitted.
-//
-// This needs to be kept in sync with the scheduler's and daemonset's fit predicates,
-// otherwise there will inevitably be pod delete create loops. This will be fixed
-// once we can extract these predicates into a common library. (#12744)
 func (kl *Kubelet) canAdmitPod(pods []*api.Pod, pod *api.Pod) (bool, string, string) {
-	if hasHostPortConflicts(pods) {
-		return false, "HostPortConflict", "cannot start the pod due to host port conflict."
-	}
-	if !kl.matchesNodeSelector(pod) {
-		return false, "NodeSelectorMismatching", "cannot be started due to node selector mismatch"
-	}
-	cpu, memory := kl.hasInsufficientfFreeResources(pods)
-	if cpu {
-		return false, "InsufficientFreeCPU", "cannot start the pod due to insufficient free CPU."
-	} else if memory {
-		return false, "InsufficientFreeMemory", "cannot be started due to insufficient free memory"
+	node, err := kl.getNodeAnyWay()
+	if err != nil {
```

@yujuhong's concern is valid here, and we might break the current users of standalone containerVM instances. But failing explicitly when failing to get node information here is desired, and I think we want to take the risk of breaking those users here.